File: algo/206_reverse_linked_list.py
"""
Reverse a singly linked list.

Example:

Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL
Follow up:

A linked list can be reversed either iteratively or recursively. Could you implement both?
"""

"""
Runtime: 20 ms, faster than 97% of Python online submissions for Reverse Linked List.
Memory Usage: 14.8 MB, less than 50% of Python online submissions for Reverse Linked List.
"""

class Solution(object):
    def reverseList(self, head):
        previous = None
        current = head

        while current:
            head = current.next
            current.next = previous
            previous = current
            current = head

        return previous

"""
Runtime: 16 ms, faster than 98.58% of Python online submissions for Reverse Linked List.
Memory Usage: 19.5 MB, less than 5.05% of Python online submissions for Reverse Linked List.
"""
class Solution(object):

    def reverseList(self, head):

        def helper(right, current):

            if not right:
                return current

            temp = right.next
            right.next = current


            return helper(temp, right)

        return helper(head, None)

File: algo/206_reverse_linked_list.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

/*
Runtime: 0ms Beats 100% of users with GoLang
Memory Usage: 4.35 MB Beats 83.03% of users with GoLang
*/
func reverseList(head *ListNode) *ListNode {
	var current *ListNode
	next := head

	for next != nil {
		tmp := next.Next
		next.Next = current

		current = next
		next = tmp
	}

	return current
}
